bugsnag-plugin-android-ndk/src/main/jni/metadata.c

```diff
@@ -40,58 +41,207 @@ void bsg_populate_metadata_value(JNIEnv *env, bugsnag_metadata *dst,
-  jni_cache->map_key_set =
-      (*env)->GetMethodID(env, jni_cache->map, "keySet", "()Ljava/util/Set;");
-  jni_cache->map_size = (*env)->GetMethodID(env, jni_cache->map, "size", "()I");
-  jni_cache->map_get = (*env)->GetMethodID(
+      bsg_safe_get_method_id(env, jni_cache->hash_map, "get",
+                             "(Ljava/lang/Object;)Ljava/lang/Object;");
+  if (jni_cache->hash_map_get == NULL) {
+    return NULL;
+  }
+
+  // lookup Map.keySet()
+  jni_cache->map_key_set = bsg_safe_get_method_id(env, jni_cache->map, "keySet",
+                                                  "()Ljava/util/Set;");
+  if (jni_cache->map_key_set == NULL) {
+    return NULL;
+  }
+
+  // lookup Map.size()
+  jni_cache->map_size =
+      bsg_safe_get_method_id(env, jni_cache->map, "size", "()I");
+  if (jni_cache->map_size == NULL) {
+    return NULL;
+  }
+
+  // lookup Map.get()
+  jni_cache->map_get = bsg_safe_get_method_id(
       env, jni_cache->map, "get", "(Ljava/lang/Object;)Ljava/lang/Object;");
+  if (jni_cache->map_get == NULL) {
+    return NULL;
+  }
+
+  // lookup com/bugsnag/android/NativeInterface
   jni_cache->native_interface =
-      (*env)->FindClass(env, "com/bugsnag/android/NativeInterface");
-  jni_cache->get_app_data = (*env)->GetStaticMethodID(
+      bsg_safe_find_class(env, "com/bugsnag/android/NativeInterface");
+  if (jni_cache->native_interface == NULL) {
+    return NULL;
+  }
+
+  // lookup NativeInterface.getApp()
+  jni_cache->get_app_data = bsg_safe_get_static_method_id(
       env, jni_cache->native_interface, "getApp", "()Ljava/util/Map;");
-  jni_cache->get_device_data = (*env)->GetStaticMethodID(
+  if (jni_cache->get_app_data == NULL) {
+    return NULL;
+  }
+
+  // lookup NativeInterface.getDevice()
+  jni_cache->get_device_data = bsg_safe_get_static_method_id(
       env, jni_cache->native_interface, "getDevice", "()Ljava/util/Map;");
-  jni_cache->get_user_data = (*env)->GetStaticMethodID(
+  if (jni_cache->get_device_data == NULL) {
+    return NULL;
+  }
+
+  // lookup NativeInterface.getUser()
+  jni_cache->get_user_data = bsg_safe_get_static_method_id(
       env, jni_cache->native_interface, "getUser", "()Ljava/util/Map;");
-  jni_cache->get_metadata = (*env)->GetStaticMethodID(
+  if (jni_cache->get_user_data == NULL) {
+    return NULL;
+  }
+
+  // lookup NativeInterface.getMetadata()
+  jni_cache->get_metadata = bsg_safe_get_static_method_id(
       env, jni_cache->native_interface, "getMetadata", "()Ljava/util/Map;");
-  jni_cache->get_context = (*env)->GetStaticMethodID(
+  if (jni_cache->get_metadata == NULL) {
+    return NULL;
+  }
+
+  // lookup NativeInterface.getContext()
+  jni_cache->get_context = bsg_safe_get_static_method_id(
       env, jni_cache->native_interface, "getContext", "()Ljava/lang/String;");
+  if (jni_cache->get_context == NULL) {
+    return NULL;
+  }
   return jni_cache;
 }
 
+// FIXME problematic global ref?
```

Comment:
What is the issue? From the looks of it, it creates a key, performs a lookup, and returns the result. It's on the caller to not attempt to store the result beyond the NDK function call.

Reply:
I don't think there is an issue - this was down to a misunderstanding that only functions marked with `JNI_EXPORT` create local references to objects.